### ISSUE TITLE
Notify mailboxes

### ISSUE THREAD

**Issue by jeremyjh** [dpc/mioco#7]
Hi, I'm contemplating  a mailbox functionality for mioco based around the notify feature of mio, but wanted to get some feedback and see if this kind of functionality and a PR is wanted before I go any further on my own.

In the reddit thread yesterday there was some discussion about capability for communicating between coroutines as well as with the outside world and I think notify-based mailboxes would accomplish this. I need functionality like this if I am to use mioco as a basis for my own (conceived) project, which is a small actors library loosely based on Haskell's distributed-process library (Cloud Haskell). 

So this mailbox functionality may seem a bit too high-level for something like mioco but let me just share my sketchy design concept for the user API and if it can be simplified thats great. 

Thanks!

Jeremy

```rust

//A unique identifier for a co-routine that can be shared with the outside world. This may
//simply be index.
struct CoID {
    id: usize
}

//Message should be Send so this will work from the outside world.
type Message = Box<Any + 'static + Send>;

struct Coroutine {
    //collection of messages which have been delivered to this co-routine
    mailbox: Vec<Message>,
..
}


//We'd need to expose this through some kind of proxy type returned
//by mioco::start I think - that is the handle to the outside world. 
//send uses mio::Sender::send to deliver the message; the Handler
//needs to implement mio::Handler::notify to put this in the correct mailbox and 
//resume the coroutine if it is blocked on receive.
fn send(coid: CoID, message: Message) -> io::Result<()> 

//This may belong to MiocoHandle...
//A coroutine can call this to attempt to receive a message of the specified type.
//Iterate through this Co's mailbox attempting downcast<T> and return
//the first success result. If no matching message is found we set a timeout and will attempt 
//the downcast again on any new message delivered before timeout expires. 
//Return None at timeout.
fn receive_timeout<T: Any+Send>(delay: u64) -> Option<Box<T>>

**Comment by dpc** [comment]
Hi,

Agreed on everything except `receive_timeout` interface. Blocking/timeout operations don't seem to blend into existing API.

It seems to me that receiving end should be just normal `TypedEventSource<Mailbox>` implementing `fn recv<T: Any+Send>(&mut self) -> Option<Box<T>>`. This way handler can block/select on Mailbox just like with every other event source. Handler could get  `TypedEventSource<Mailbox>` doing a `wrap()` on `Mailbox` type that would register it. Would that work? This type of API has a nice property, that `Mailbox` can be retired in the future and replaced with 

How would sharing of coroutine IDs with an outside world work? I agree that `mioco::start` should return some "control structure" that outside world can use, but as coroutines can be created in any number at runtime, they could get any ID. Maybe mailboxes should be pairs of objects created before `mioco::start` and then passed around. This way no coroutines IDs are needed, and whichever coroutine do `MiocoHandle::wrap` on a receiving-end is the receiving target to the send-end. This way even multicast operations could be potentially supported if receiving end can be `clone()`-d.

Could such Mailbox be working both ways somehow (maybe not initially, but at least conceptually, to be implement later).

Anyway, thanks for bringing this subject up and giving it some thought! Let me know what you think.

**Comment by jeremyjh** [comment]
Hi, I think I could use that to do what I want and it does sound like it fits better into the current design. I am not sure how the implementation works; I think Mailbox has to be mio::Evented and it looks like that requires a file descriptor in the Selector.  On Linux maybe signalfd would be a good way to create that fd; I'm not sure about other platforms. The fd requirement is why I was looking into send/notify alternative.

**Comment by dpc** [comment]
It seems to me that `mioco::Evented` requiring `mio::Evented` might be possible to be relaxed, moving `reregister`, `deregister` etc. as `mio::Evented` methods. This way types that are not `mio::Evented` could do something different if are not `mio::Evented`. Or just stuffing an `enum` somewhere doing different things for `mio::Evented` and different for internal mioco-provided types, would work too. Traits-ways would be preferable, as it might be possible to have 3rd party mioco communication channel types etc.

**Comment by jeremyjh** [comment]
Yes I think that would work. If I can make a TypedEventSource<Mailbox>, register it and have some way for it to Signal when it is readable then probably the rest of what I want to do can be implemented outside mioco itself.

**Comment by dpc** [comment]
OK. Please keep me (and potentially others) in loop on gitter about it! I've seen that this has been an most requested feature, so we can coordinate.

**Comment by jeremyjh** [comment]
Fixed by commit 24ec5661e173b10026157ede1245c12c056d003c
